dvc/external_repo.py

```diff
@@ -218,58 +251,13 @@ def _get_tree_for(self, repo, **kwargs):
             kw["fetch"] = True
         return RepoTree(repo, **kw)
 
-    @staticmethod
-    def _fix_local_remote(orig_repo, src_repo, remote_name):
-        # If a remote URL is relative to the source repo,
-        # it will have changed upon config load and made
-        # relative to this new repo. Restore the old one here.
-        new_remote = orig_repo.config["remote"][remote_name]
-        old_remote = src_repo.config["remote"][remote_name]
-        if new_remote["url"] != old_remote["url"]:
-            new_remote["url"] = old_remote["url"]
-
-    @staticmethod
-    def _add_upstream(orig_repo, src_repo):
-        # Fill the empty upstream entry with a new remote pointing to the
-        # original repo's cache location.
-        cache_dir = src_repo.cache.local.cache_dir
-        orig_repo.config["remote"]["auto-generated-upstream"] = {
-            "url": cache_dir
-        }
-        orig_repo.config["core"]["remote"] = "auto-generated-upstream"
-
     def make_repo(self, path):
-        repo = Repo(path, scm=self.scm, rev=self.get_rev())
-
-        self._setup_cache(repo)
-        self._fix_upstream(repo)
-
-        return repo
-
-    def _setup_cache(self, repo):
-        repo.config["cache"]["dir"] = self.cache_dir
-        repo.cache = Cache(repo)
-        if self.cache_types:
-            repo.cache.local.cache_types = self.cache_types
-
-    def _fix_upstream(self, repo):
-        if not os.path.isdir(self.url):
-            return
-
-        try:
-            rel_path = os.path.relpath(repo.root_dir, self.root_dir)
-            src_repo = Repo(PathInfo(self.url) / rel_path)
-        except NotDvcRepoError:
-            return
-
-        try:
-            remote_name = repo.config["core"].get("remote")
-            if remote_name:
-                self._fix_local_remote(repo, src_repo, remote_name)
-            else:
-                self._add_upstream(repo, src_repo)
-        finally:
-            src_repo.close()
+        config = self._cache_config.copy()
+        if os.path.isdir(self.url):
+            rel = os.path.relpath(path, self.root_dir)
+            repo_path = os.path.join(self.url, rel)
+            config.update(_get_remote_config(repo_path))
+        return Repo(path, scm=self.scm, rev=self.get_rev(), config=config)
 
     @wrap_with(threading.Lock())
     def _get_cache_dir(self):
```

Comment:
This should be out of this class, it is more about the url caching mechanism than ExternalRepo itself. Will likely get revisited when introducing global persistent git repo cache.